devcareer/code-jammers-backend
# #175218188 Newsletter and subscription APIs

## What does this PR do ##
- This PR enables users to subscribe or unsubscribe from a newsletter.
- It enables only Users with an Admin or Super Admin role create a newsletter.
- Subscribers receive newsletters immediately they're sent out

## Description of tasks to be completed? ##
- Create `Newsletter` and `Subscriber` models and migrations. Both models will have a `belongsToMany` association with each other and a `Newsletter_Subscriber` model as a through/join table. The `Newsletter_Subscriber` will have a `belongTo` association with both Newsletter and Subscriber models.
- When a subscriber is created, a verification mail will be sent immediately to the provided email address for the subscriber to verify they own the email address. If the user clicks the `Verify Me` button, the verified column in the subscriber database gets updated to `true`.
- Make use of isAdmin middleware to ensure only Admin or Super Admins can create a newsletter.
- Integrate Sendgrid to help send mails to provided email addresses
-  When a newsletter is created, it is sent to the email addresses of subscribers immediately. It also saves the newsletter to the Newsletter database. The Subscriber gets updated with the title of each newsletter received by the subscriber. 
- Subscribers can unsubscribe from the newsletter


## How should this be manually tested/checked? ##
1. Clone the PR
2. Cd into the branch ft-user-can-subscribe/unsubsrcibe-to-newsletter-175218188
3. Set up database connection string and other secrets as shown in the `.env.sample` file.
4. Install dependencies using `npm install`.
5. Run npm run migrate to create the database tables
6. Run npm run seed to feed the database
7. To test all endpoints, `npm run test` comes handy.
8. Start the server by running `npm run start-dev`.
9. Subscribe a user by making a `POST` request to `http://localhost:3000/api/v1/newsletter/subscribe` with the details similar to

> {

       "firstName": "Jerry",
       "email": "yourValidEmailAddress@gmail.com"

>   }

10. Check your mail box and open the verification mail from `Know Africa`. Click on the `Verify Me` button to confirm your mail.
 
11. Sign in a user with admin role to create a newsletter via `http://localhost:3000/api/v1/users/signin` using a `POST` method with
          

> {

    "email: "ufuoma@gmail.com",
    "password": 12345

> }

and copy the token (without the quotes) from the response body.
12. In a new tab, under the Authorization tab, change the TYPE to Bearer Token and paste the token in the space available. Make a POST request to `http://localhost:3000/api/v1/newsletter/admin/create_newsletter` to create a newsletter. To create a newsletter, use the format below
          

>  {

             "title": "What is Lorem Ipsum?",
             "message": "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the 
                        industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and 
                        scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into 
                        electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of 
                        Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like 
                         Aldus PageMaker including versions of Lorem Ipsum."
         

>  }

13. Check your registered email address to see the just created newsletter in your mail box.
14. To Unsubscribe from the newsletter, make a `GET` request to `http://localhost:3000/api/v1/newsletter/unsubscribe` with details similar to 
          

> {

           "email": "yourValidEmailAddress@gmail.com"

> }

15. Create a new newsletter following step 12 above. You shouldn't see the latest newsletter in your mail box.
## Architectural flowchart: ##
[Subscribe flowchart](https://docs.google.com/document/d/1C4oPY4V8j8N3-ChQ-rVjKN-VLm_qYEWVMtboOI9cQDs/edit?usp=sharing)
[Unsubscribe flowchart](https://docs.google.com/document/d/1s2JLggTDrdK6MUv-E2_QxOCkumnRrZN7M1WzFD9iVCM/edit?usp=sharing)
